pyrealm/demography/canopy_functions.py

```diff
@@ -0,0 +1,411 @@
+        crown_area: Crown area of each cohort
+        stem_height: Stem height of each cohort
+        m: Canopy shape parameter ``m``` for each cohort
+        n: Canopy shape parameter ``n``` for each cohort
+        q_m: Canopy shape parameter ``q_m``` for each cohort
+        z_max: Canopy shape parameter ``z_m``` for each cohort
+        target_area: A target projected crown area.
+        validate: Boolean flag to suppress argument validation.
+    """
+    # Convert z to array for validation and typing
+    z_arr = np.array(z)
+
+    if validate:
+        _validate_z_args(
+            z_arr, n_individuals, crown_area, stem_height, m, n, q_m, z_max
+        )
+
+    q_z = calculate_relative_canopy_radius_at_z(
+        z=z_arr, stem_height=stem_height, m=m, n=n, validate=False
+    )
+    # Calculate A(p) for the stems in each cohort
+    A_p = calculate_stem_projected_crown_area_at_z(
+        z=z_arr,
+        q_z=q_z,
+        stem_height=stem_height,
+        crown_area=crown_area,
+        q_m=q_m,
+        z_max=z_max,
+        validate=False,
+    )
+
+    return (A_p * n_individuals).sum() - target_area
+
+
+def calculate_stem_projected_leaf_area_at_z(
+    z: NDArray[np.float32],
+    q_z: NDArray[np.float32],
+    stem_height: NDArray[np.float32],
+    crown_area: NDArray[np.float32],
+    f_g: NDArray[np.float32],
+    q_m: NDArray[np.float32],
+    z_max: NDArray[np.float32],
+    validate: bool = True,
+) -> NDArray[np.float32]:
+    """Calculate projected leaf area above a given height.
+
+    This function calculates the projected leaf area of a set of stems with given
+    properties at a set of vertical heights. This differs from crown area in allowing
+    for crown openness within the crown of an individual stem that results in the
+    displacement of leaf area further down into the canopy. The degree of openness is
+    controlled by the crown gap fraction property of each stem.
+
+    The stem properties are given in the arguments
+    ``stem_height``,``crown_area``,``f_g``,``q_m`` and ``z_max``, which must be
+    one-dimensional arrays ('row vectors') of equal length. The array of vertical
+    heights ``z`` accepts a range of input shapes (see
+    :meth:`~pyrealm.demography.canopy_functions.calculate_relative_canopy_radius_at_z`
+    ) and this function then also requires the expected relative stem radius (``q_z``)
+    calculated from those heights.
+
+    Args:
+        z: Vertical heights on the z axis.
+        q_z: Relative crown radius at heights in z.
+        crown_area: Crown area for a stem
+        stem_height: Total height of a stem
+        f_g: Within crown gap fraction for each stem.
+        q_m: Canopy shape parameter ``q_m``` for each stem
+        z_max: Height of maximum canopy radius for each stem
+        validate: Boolean flag to suppress argument validation.
+    """
+
+    # TODO - could this merge with the stem crown area function? A lot of overlap, so
```

Comment:
Again, should be an issue

Reply:
This one is a bad idea on reflection - the crown area function will get run _a lot_ from within the solver, so keeping it as lean as possible makes a lot of sense.